src/base/utils/os.cpp

```diff
@@ -0,0 +1,412 @@
+                        qDebug("Found python.exe at %s", qUtf8Printable(path));
+                        found = true;
+                    }
+                }
+            }
+
+            if (!found)
+                path = QString();
+
+            ::RegCloseKey(hkPythonCore);
+        }
+
+        return path;
+    }
+
+    QString getAssoc(const QString &cls)
+    {
+        return (QSettings {"HKEY_CLASSES_ROOT", QSettings::NativeFormat}).value(cls + "/Default").toString();
+    }
+
+    void setAssoc(const QString &cls, const QString &value)
+    {
+        QSettings reg {"HKEY_CURRENT_USER\\Software\\Classes", QSettings::NativeFormat};
+        const QString oldProgID = reg.value(cls + "/Default").toString();
+        if (!oldProgID.isEmpty() && (oldProgID != value))
+            reg.setValue(cls + "/OpenWithProgids/" + oldProgID, "");
+        reg.setValue(cls + "/Default", value);
+        ::SHChangeNotify(SHCNE_ASSOCCHANGED, SHCNF_IDLIST, 0, 0);
+    }
+}
+
+QString Utils::OS::getPythonPath()
+{
+    QString path = pythonSearchReg(USER);
+    if (!path.isEmpty())
+        return path;
+
+    path = pythonSearchReg(SYSTEM_32BIT);
+    if (!path.isEmpty())
+        return path;
+
+    path = pythonSearchReg(SYSTEM_64BIT);
+    if (!path.isEmpty())
+        return path;
+
+    // Fallback: Detect python from default locations
+    const QStringList dirs = QDir("C:/").entryList(QStringList("Python*"), QDir::Dirs, QDir::Name | QDir::Reversed);
+    foreach (const QString &dir, dirs) {
+        const QString path("C:/" + dir + "/");
+        if (QFile::exists(path + "python.exe"))
+            return path;
+    }
+
+    return QString();
+}
+
+bool Utils::OS::areAssociationsSet() noexcept
+{
+#ifndef WIN_HAS_DEFAULT_PROGRAMS_API
+    Q_ASSERT(QSysInfo::WindowsVersion < QSysInfo::WV_VISTA);
+    return (isTorrentFileAssocSet() && isMagnetLinkAssocSet());
+#else
+    Q_ASSERT(QSysInfo::WindowsVersion < QSysInfo::WV_WINDOWS8);
+
+    if (QSysInfo::WindowsVersion < QSysInfo::WV_VISTA)
+        return (isTorrentFileAssocSet() && isMagnetLinkAssocSet());
+
+    // Vista, Win7
+    BOOL isDefault = FALSE;
+    try {
+        const COMLib comLib {COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE};
+
+
+        auto pAssoc = comLib.createObject<IApplicationAssociationRegistration>(
+                          CLSID_ApplicationAssociationRegistration, nullptr, CLSCTX_INPROC);
+        if (pAssoc)
+            pAssoc->QueryAppIsDefaultAll(AL_EFFECTIVE, APP_REGISTRY_NAME, &isDefault);
+    }
+    catch (const std::runtime_error &) {}
+
+    return (isDefault == TRUE);
+#endif
+}
+
+void Utils::OS::setAssociations()
+{
+#ifndef WIN_HAS_DEFAULT_PROGRAMS_API
+    Q_ASSERT(QSysInfo::WindowsVersion < QSysInfo::WV_VISTA);
+
+    setTorrentFileAssoc(true);
+    setMagnetLinkAssoc(true);
+#else
+    Q_ASSERT(QSysInfo::WindowsVersion < QSysInfo::WV_WINDOWS8);
+
+    if (QSysInfo::WindowsVersion < QSysInfo::WV_VISTA) {
```

Comment:
Why does the condition repeat assertion in line 283?

Reply:
Typo. `>=` is supposed to be here.